components/brave_ads/browser/ads_p2a.cc

```diff
@@ -112,21 +112,25 @@ void RecordInWeeklyStorageAndEmitP2AHistogramAnswer(
 
 void EmitP2AHistogramAnswer(
     const std::string& name,
-    int count_value) {
-  const uint16_t* it =
-      std::lower_bound(kBinIntervals, std::end(kBinIntervals), count_value);
-  const uint16_t answer = it - kBinIntervals;
+    uint16_t count_value) {
+  const uint16_t* iter = std::lower_bound(kIntervalBuckets,
+      std::end(kIntervalBuckets), count_value);
+  const uint16_t bucket = iter - kIntervalBuckets;
 
-  for (const char* question_name : kP2AQuestionList) {
-    if (name == question_name) {
-      base::UmaHistogramExactLinear(question_name, answer, 8);
+  for (const char* question_name : kP2AQuestionNameList) {
+    if (name != question_name) {
+      continue;
     }
+
+    base::UmaHistogramExactLinear(question_name, bucket,
+        base::size(kIntervalBuckets) + 1);
   }
 }
 
 void SuspendP2AHistograms() {
-  for (const char* question_name : kP2AQuestionList) {
-      base::UmaHistogramExactLinear(question_name, INT_MAX, 8);
+  for (const char* question_name : kP2AQuestionNameList) {
+    base::UmaHistogramExactLinear(question_name, INT_MAX,
```

Comment:
@moritzhaller I would also add a DCHECK to `DirectEncodingProtocol::Perturb` 

basically `UmaHistogramExactLinear` expectes maximum value, not the count of possible values, so strictly speaking, we'd better keep `base::size(kIntervalBuckets)` here and fix `BraveP3AService::OnHistogramChanged` - for some reason it decrements bucket count before calling `Perturb`.

Reply:
@iefremov thinking about this again to make sure we're not counting wrong:

- because we use `lower_bound()` on `{0, 5, 10, 20, 50, 100, 250, 500 }` we have 9 possible values for `bucket`: `0|1-5|6-10|11-20|21-50|51-100|101-250|251-500|501+` => `0, 1, ... ,8`
- [LinearHistogram](https://source.chromium.org/chromium/chromium/src/+/master:base/metrics/histogram.h;l=12;drc=d81c5852498699fe3cd812e78d31c77c28e29281;bpv=1;bpt=1) always has `0` as a min value and will create an additional overflow bucket, so `base::size(kIntervalBuckets) + 1` will result in an overflow bucket "9" and increase the bucket count to 10. Which is why we need to subtract when using the count for perturbation (we handle "overflow" already with our "501+" interval so we don't care about the overflow bucket and it should never be sent even after randomisation. We always want to choose from buckets 0-8 and never from bucket 9)
- Now I see a new issue, since we would never report a 0 bucket when reporting the true answer since we always increment when reading from `WeeklyStorage` before emitting a histogram. But the perturbation step allows for sending the 0 bucket. So we should probably either initialise weekly storage for all questions with 0 so we send true 0 counts, or remove it from the perturbation step. I'd prefer the former.

@tmancey @iefremov wdyt? Would it make sense to initialise `WeeklyStorage` with zero counts for all questions?